src/internal/internal_genorm.cc

```diff
@@ -463,62 +438,51 @@ void norm(
-                lda[q] = 0;
-                mb[q] = A.tileMb(irange[q][0]);
-                nb[q] = A.tileNb(jrange[q][0]);
-                for (int64_t i = irange[q][0]; i < irange[q][1]; ++i) {
-                    for (int64_t j = jrange[q][0]; j < jrange[q][1]; ++j) {
-                        if (A.tileIsLocal(i, j) &&
-                            device == A.tileDevice(i, j))
-                        {
-                            a_host_array[batch_count] = A(i, j, device).data();
-                            lda[q] = A(i, j, device).stride();
-                            ++group_count[q];
-                            ++batch_count;
-                        }
-                    }
-                }
-            }
+            int64_t batch_size = A_tiles_set.size();
+            scalar_t** a_array_host = A.array_host( device, queue_index );
+
+            auto group_params = device_regions_build<false, 1, scalar_t>(
+                    {A},
+                    {a_array_host},
+                    device,
+                    {},
+                    irange, jrange );
 
-            real_t* vals_host_array = vals_host_arrays[device].data();
-            real_t* vals_dev_array = vals_dev_arrays[device];
+            scalar_t** a_array_dev = A.array_device(device, queue_index);
+
+            vals_host_arrays[ device ].resize( batch_size*ldv );
+            real_t* vals_host_array = vals_host_arrays[ device ].data();
+            blas::Queue* queue = A.compute_queue( device, queue_index );
+            real_t* vals_dev_array = blas::device_malloc<real_t>( batch_size*ldv, *queue );
 
             // Batched call to compute partial results for each tile.
             {
                 trace::Block trace_block("slate::device::genorm");
 
-                blas::Queue* queue = A.compute_queue(device, queue_index);
-
-                blas::device_memcpy<scalar_t*>(a_dev_array, a_host_array,
-                                    batch_count,
-                                    blas::MemcpyKind::HostToDevice,
-                                    *queue);
-
-                for (int q = 0; q < 4; ++q) {
-                    if (group_count[q] > 0) {
-                        device::genorm(in_norm, scope,
-                                       mb[q], nb[q],
-                                       a_dev_array, lda[q],
-                                       vals_dev_array, ldv,
-                                       group_count[q], *queue);
-                        a_dev_array += group_count[q];
-                        vals_dev_array += group_count[q] * ldv;
-                    }
-                }
 
-                vals_dev_array = vals_dev_arrays[device];
+                blas::device_memcpy<scalar_t*>(a_array_dev, a_array_host,
+                                               batch_size,
+                                               blas::MemcpyKind::HostToDevice,
+                                               *queue);
```

Comment:
Don't need `blas::MemcpyKind::HostToDevice,`. That usage is actually [marked as deprecated](https://github.com/icl-utk-edu/blaspp/blob/master/include/blas/device.hh#L629-L644) in Doxygen, but not yet marked deprecated using C++ `[[deprecated]]`.